---
title: "hiv_intsite_compiled"
author: "Vincent Wu"
date: "10/3/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# load pkgs
```{r, warning=FALSE, message=FALSE}
library(ggplot2)
library(tidyverse)
library(glue)
library(GenomicRanges)
library(hiAnnotator)
library(RColorBrewer)
library(scales)
library(circlize)
library(grid)
library(gridExtra)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
library(org.Hs.eg.db)
library(ChIPseeker)
library(BSgenome.Hsapiens.UCSC.hg38)
library(hotROCs)
library(reshape2)
library(svglite)
library(egg)

# known overlapping functions
trim <- GenomicRanges::trim

source("scripts/mrc_generator.R")
source("scripts/heatmapMaker.R")
source("scripts/epigeneticFeatures.R")
source("scripts/random_site.R")
source("scripts/heatmap_to_ggplot2.R")
source("scripts/hiv_int_helper_fx.R")
```

# load and merge data files
```{r}
df_compiled <- readRDS("data/df_compiled.rds")

df_compiled <- df_compiled %>%
  dplyr::filter(patient_anon != "") %>%
  dplyr::filter(seqnames != "chrM") %>%
  mutate(exp = str_match(run, pattern = "(\\w+)_(\\d)")[, 2]) %>%
  mutate(exp = ifelse(exp == "art", "ART", str_to_title(exp)),
         cell_oldname = cell)

# set proper names for metacolumns
df_compiled$location <- sapply(df_compiled$location, function(x) {
  val <- ""
  if (x == "pbmc") {
    val <- "PBMC"
  } else if (x == "tonsil") {
    val <- "Tonsil"
  } else {
    val <- toupper(x)
  }
  
  return(val)
})

df_compiled$cell <- sapply(seq_along(df_compiled$cell_oldname), function(i) {
  row <- df_compiled[i, ]
  cell <- row$cell_oldname
  loc <- row$location
  exp <- row$exp
  
  val <- ""
  error <- "error"
  
  if (exp == "ART" & loc == "Tonsil") {
    val <- case_when(
      cell == "naive" ~ "Naive",
      cell == "69pos" ~ "HLA-DR- CD69+",
      cell == "69neg" ~ "HLA-DR- CD69-",
      cell == "gctfh" ~ "HLA-DR- GC-Tfh",
      TRUE ~ error
    )
  } else if (exp == "ART" & loc == "PBMC") {
    val <- case_when(
      cell == "naive" ~ "Naive",
      cell == "tcm" ~ "CD69- Tcm",
      cell == "tem" ~ "CD69- Tem",
      TRUE ~ error
    )
  } else {
    val <- case_when(
      cell == "naive" ~ "Naive",
      cell == "69pos" ~ "CD69+",
      cell == "69neg" ~ "CD69-",
      cell == "gctfh" ~ "GC-Tfh",
      cell == "tcm" ~ "Tcm",
      cell == "tem" ~ "Tem",
      cell == "temra" ~ "Temra",
      TRUE ~ error
    )    
  }
  
  return(val)
})

# label in/out of gene
df_compiled <- df_compiled %>%
  mutate(gene_plot_noid = case_when(
    in_gene == "FALSE" & abs(nearest_geneDist) <= 50000 ~ glue("{nearest_gene}+"),
    in_gene == "FALSE" & abs(nearest_geneDist) > 50000 ~ glue("{nearest_gene}#"),
    TRUE ~ in_gene
  )) %>%
  mutate(gene_plot = paste(gene_plot_noid, posID))

# add new numbering system
df_compiled <- df_compiled %>%
  mutate(patient_anon_v2 = case_when(
    patient_anon == "A" ~ "T1",
    patient_anon == "B" ~ "T2",
    patient_anon == "C" ~ "T3",
    patient_anon == "A_D" ~ "A1",
    patient_anon == "A_E" ~ "A2",
    patient_anon == "A_F" ~ "A3",
    patient_anon == "C_G" ~ "C1",
    patient_anon == "C_H" ~ "C2",
    patient_anon == "C_I" ~ "C3",
    TRUE ~ "Error"
  ))

gr_compiled <- makeGRangesFromDataFrame(df_compiled,
                                        keep.extra.columns = TRUE)
```

# stat calculations
```{r}
df_compiled %>%
  group_by(exp, patient_anon_v2, location, cell) %>%
  summarise(total_abund = sum(abund)) %>%
  arrange(desc(total_abund)) %>% 
  ungroup() %>%
  group_by(exp, patient_anon_v2, location) %>%
  mutate(location_sum = sum(total_abund)) %>%
  ungroup() %>%
  group_by(exp, patient_anon_v2) %>%
  mutate(patient_sum = sum(total_abund)) %>%
  arrange(exp, desc(patient_sum), desc(location_sum)) %>%
  write.table("df_compiled_grouped_stats.csv", col.names = TRUE,
              row.names = FALSE, quote = FALSE, sep = "\t")
```

# in/out of gene
```{r}
df_compiled %>%
  ungroup() %>%
  mutate(feature = ifelse(in_gene == "FALSE", "Outside", "In")) %>%
  group_by(exp, location, cell) %>%
  mutate(group_n = sum(abund)) %>%
  group_by(exp, location, cell, feature) %>%
  summarise(percent_feature = sum(abund) / max(group_n) * 100,
            sum_feature = sum(abund),
            group_abund = max(group_n)) %>%
  mutate(x_axis = paste0(gsub(" ", "\n", cell),
                         "\n(",
                         group_abund,
                         ")")) %>%
  ungroup() %>%
  mutate(exp = factor(exp, levels = c("Acute", "Chronic", "ART")),
         location = factor(location, levels = c("PBMC", "CLN", "ILN", "Tonsil"))) %>%
  ggplot(aes(x = x_axis, y = percent_feature, fill = feature)) +
  geom_col(position = position_dodge2(preserve = "single")) +
  labs(fill = "Feature",
       y = "% sonic abundance",
       x = "Cell subset\n(total sonic abundance)") +
  theme_classic() +
  theme(strip.background = element_rect(fill = "#FFFFFF", colour = "#FFFFFF"),
        strip.text = element_text(face = "bold", size = 10),
        legend.title = element_text(size = 10),
        legend.text = element_text(size = 8),
        axis.text.x = element_text(size = 8)) +
  guides(color = guide_legend(override.aes = list(size = 0.75))) +
  facet_wrap(~ exp + location,
             scales = "free_x",
             labeller = facet_label_clean)

ggsave(filename = "figs/inout_gene_subset.svg", device = "svg", width = 9, height = 6)
```


# temp explore
```{r}
# checking overlap
df_compiled %>% 
  group_by(patient_anon, posID) %>%
  filter(n() > 1) %>%
  arrange(posID, .by_group = TRUE) %>%
  summarise(count = sum(n()))
```


# patient-wide simulations
```{r}
calc_overlap <- function(ids, bin_index) {
  abund_ids <- ids[duplicated(ids)]
  total_count <- 0
  
  for (i in abund_ids) {
    bin_count <- 0
    for (b in seq(2, length(bin_index) + 1, 1)) {
      start <- bin_index[b - 1]
      
      if (b == length(bin_index) + 1) {
        end <- length(ids)
      } else {
        end <- bin_index[b] - 1        
      }
      
      bin <- ids[start:end]
      if (sum(bin == i) > 0) {
        bin_count <- bin_count + 1
      }
    }
    
    if (bin_count >= 2) {
      total_count <- total_count + 1
    }
  }
  
  return(total_count)
}

uniq_patients <- unique(df_compiled$patient_anon)
graphs <- list()

# p <- "B"

graphs <- list()
for (p in uniq_patients) {
  print(paste("Running simulation for", p))
  
  trial_df <- df_compiled %>%
    filter(patient_anon == p) %>%
    mutate(location_cell = paste(location, cell, sep = "_")) %>%
    mutate(group_id = group_indices(., location_cell)) %>%
    arrange(group_id)
  
  bin_index <- which(!duplicated(trial_df$group_id))
  ids <- trial_df$posID
  weights <- trial_df$abund
  
  # make pool with correct abundances
  pool_ids <- rep(ids, times = floor(weights))
  
  # find weights by group
  g_sum_df <- trial_df %>%
    group_by(group_id) %>%
    summarise(g_sum = sum(floor(abund)))
  
  weights_grouped <- g_sum_df$g_sum
  
  bin_index_exp <- cumsum(weights_grouped)
  bin_index_exp <- bin_index_exp[-length(bin_index_exp)] + 1
  bin_index_exp <- c(1, bin_index_exp)
  
  # calculate overlap
  act_overlaps <- sum(duplicated(ids))
  
  # run simulations for overlap
  test <- sapply(seq(1, 10000), function(x) {
    pool_scramble <- sample(pool_ids)
    calc_overlap(pool_scramble, bin_index_exp)
  })
  
  # weights_grouped
  
  print(paste("Actual overlaps:", act_overlaps))

  # clone_info <- trial_df %>% filter(abund > 1)
  # print(summary(clone_info$abund))
  
  g_df <-  data.frame(x = test)
  g <- ggplot(data = g_df, aes(x = x)) +
    geom_vline(xintercept = act_overlaps, color = "#AAAAAA", linetype = "dashed") +
    stat_ecdf(geom = "step") +
    expand_limits(x = 0) +
    # geom_histogram(fill = "#BBBBBB", color = "#AAAAAA", bins = 5) +
    # geom_density(fill = "#CCCCCC", color = "#AAAAAA") +

    labs(x = "",
         y = "",
         title = p) + 
    theme_classic()
  
  graphs[[p]] <- g
    
  ggsave(filename = paste("figs/global_overlap_ecdf_", p, ".png", sep = ""),
         plot = graphs[[p]],
         device = "png",
         width = 4,
         height = 4)
}

# multiplot(plotlist = graphs, cols = 2)
```

# hotspot graph
```{r}
# TODO arrange by patient.

uniq_exp <- unique(df_compiled$exp)
hotspot_graphs <- list()
for (u in uniq_exp) {
  df_plot_all <- df_compiled %>%
    dplyr::filter(exp == u) %>%
    dplyr::filter(!(in_gene == "FALSE" & abs(nearest_geneDist <= 50000))) %>%
    arrange(desc(patient_anon_v2), desc(location), desc(cell)) %>%
    mutate(cell_patient = paste(patient_anon_v2, location, cell, sep = " / "),
           cell_patient = factor(cell_patient),
           nearest_gene = factor(nearest_gene, levels = unique(nearest_gene)))
  
  sim_matrix <- matrix(0L, nrow = length(levels(df_plot_all$cell_patient)),
                       ncol = length(levels(df_plot_all$nearest_gene)))
  
  for (i in seq_along(df_plot_all$cell_patient)) {
    df_row <- df_plot_all[i, ]
    
    row <- as.numeric(df_row$cell_patient)
    col <- as.numeric(df_row$nearest_gene)
    
    if (sim_matrix[row, col] == 0) {
      sim_matrix[row, col] <- sim_matrix[row, col] + 1
    }
  }
  
  melted_sim <- melt(sim_matrix) %>%
    mutate(Var1 = factor(Var1, labels = levels(df_plot_all$cell_patient))) %>%
    separate(Var1,
             into = c("patient", "sample_type", "cell"),
             remove = FALSE, sep = " / ") %>%
    dplyr::filter(value > 0) %>%
    mutate(Var2 = factor(Var2))
  
  cs <- base::colSums(sim_matrix)
  
  # element_text(angle = 45, vjust = 1, hjust = 1, size = 7),
  heatmap_theme <- theme(axis.text.x = element_text(size = 7, lineheight = 2),
                         axis.title.x = element_blank(),
                         panel.background = element_rect(fill = "#132b43"),
                         panel.grid.major.y = element_blank(),
                         panel.grid.minor.y = element_blank(),
                         panel.grid.major.x = element_blank(),
                         legend.position = "none")
  
  if (u == "Chronic") {
    col_of_interest <- which(colSums(sim_matrix) >= 4)
  } else {
    col_of_interest <- which(colSums(sim_matrix) >= 3)
  }

  of_interest <- melted_sim %>%
    dplyr::filter(Var2 %in% col_of_interest) %>%
    mutate(cell = gsub(" ", "\n", cell),
           x_lbl = glue("{patient}\n{sample_type}\n{cell}"))
  
  g <- ggplot(of_interest, aes(x_lbl, factor(Var2), fill = value)) +
    geom_tile(fill = "#56b1f7", width = 1, height = 1) +
    scale_y_discrete(breaks = col_of_interest,
                     labels = levels(df_plot_all$nearest_gene)[col_of_interest]) +
    theme_bw() +
    labs(y = "Gene or Nearest Gene") +
    heatmap_theme +
    coord_equal()
  
  panel_scale <- 1
  p <- egg::set_panel_size(g, height=unit(length(col_of_interest) / panel_scale, "cm"),
                          width=unit(length(unique(of_interest$Var1)) / panel_scale, "cm") )
  
  hotspot_graphs[[u]] <- p
}

cowplot::plot_grid(hotspot_graphs[["Acute"]],
                   hotspot_graphs[["ART"]],
                   hotspot_graphs[["Chronic"]], nrow=1)

ggsave(paste0("figs/", "overlap_hotspot_final.svg"),
       device = "svg", width = 70, height = 40, units = "cm")
```


# clonal expansion graph
```{r}
make_clonal_expansion_graph <- function(df, p, exp_type) {
  if (exp_type == "Acute") {
    abund_filter <- 3
  } else {
    abund_filter <- 2
  }
  
  df_rel <- df %>%
    dplyr::filter(patient_anon == p) %>%
    ungroup() %>%
    group_by(main_id, cell) %>%
    mutate(group_abund = sum(abund),
           rel_abund = abund / group_abund,
           gene_plot_2 = ifelse(abund < abund_filter, "Low Abundance", gene_plot),
           plot_lbl = ifelse(gene_plot_2 != "Low Abundance" & rel_abund >= 0.1,
                             gene_plot_noid,
                             ""),
           x_axis = paste0(location, "\n",
                          gsub(" ", "\n", cell),
                          "\n",
                          "(", sum(abund), ")")) %>%
    arrange(rel_abund, .by_group = TRUE)
  
  uniq_genes <- unique(df_rel$gene_plot_2)
    
  df_rel$gene_plot_2 <- factor(df_rel$gene_plot_2,
                                  levels = c("Low Abundance",
                                             uniq_genes[uniq_genes != "Low Abundance"]))
    
  colors <- hue_pal()(length(uniq_genes))
  colors[1] <- "#dddddd"
  
  legend_items <- length(unique(df_rel$gene_plot_2))
  legend_ncol <- legend_items %/% 10 + 1
  
  g <- ggplot(df_rel, aes(x = x_axis, y = rel_abund, fill = gene_plot_2)) +
    geom_bar(stat = "identity") +
    geom_text(aes(label = plot_lbl), size = 2.5, position = position_stack(vjust = 0.5)) +
    theme_classic() +
    theme(panel.grid.major.x = element_blank(),
          panel.grid.minor.x = element_blank(),
          legend.title = element_blank(),
          legend.text = element_text(size = 6),
          legend.position = "right") +
    guides(fill = guide_legend(override.aes = list(size = 1),
                           ncol = legend_ncol)) +
    scale_fill_manual(values = colors) +
    labs(title = p,
         x = "Patient and Cell Type (n = total inferred cells)",
         y = paste0("Relative sonic abundance \n",
         "(Low Abundance < ", abund_filter, " sonic abund)"))
  
  scaled_g <- egg::set_panel_size(g, height=unit(3, "cm"),
                          width=unit(length(unique(df_rel$x_axis)) + 3.5, "cm"))
  
  return(scaled_g)
  # ggsave(paste0("figs/", "rel_abund_", p, ".png"), device = "png", bg = "transparent")
}

uniq_patients <- unique(df_compiled$patient_anon)
clonal_graphs <- list()
for (p in uniq_patients) {
  if (startsWith(p, "A_")) {
    clonal_graphs[[p]] <- make_clonal_expansion_graph(df_compiled, p, "Acute")
  } else if (startsWith(p, "C_")) {
    clonal_graphs[[p]] <- make_clonal_expansion_graph(df_compiled, p, "Chronic")
  } else {
    clonal_graphs[[p]] <- make_clonal_expansion_graph(df_compiled, p, "ART")
  }
}

art_rel_abund <- cowplot::plot_grid(clonal_graphs[["A"]], clonal_graphs[["B"]], clonal_graphs[["C"]], ncol = 1, align = "h")
cowplot::save_plot("figs/rel_abund_art.svg", art_rel_abund, base_width = NULL, base_height = 25, device = "svg")

acute_rel_abund <- cowplot::plot_grid(clonal_graphs[["A_D"]], clonal_graphs[["A_E"]], clonal_graphs[["A_F"]], ncol = 1, align = "h")
cowplot::save_plot("figs/rel_abund_acute.svg",
                   acute_rel_abund, base_width = NULL, base_height = 25, device = "svg")

chronic_rel_abund <- cowplot::plot_grid(clonal_graphs[["C_G"]], clonal_graphs[["C_H"]], clonal_graphs[["C_I"]], ncol = 1, align = "h")
cowplot::save_plot("figs/rel_abund_chronic.svg",
                   chronic_rel_abund, base_width = NULL,
                   base_height = 25, device = "svg")

# 
# ggarrange(clonal_graphs[["A"]], clonal_graphs[["B"]], nrow=1)
```


# circos plot
```{r}
uniq_patients <- unique(df_compiled$patient_anon)
# uniq_cells <- unique(paste(df_compiled$location, df_compiled$cell, sep = "_"))
# cell_colors <- brewer.pal(length(uniq_cells), "Paired")

df_uniq_cells <- df_compiled %>%
  mutate(sample_cell = paste(location, cell, sep = "_")) %>%
  group_by(exp) %>%
  distinct(sample_cell)

df_uniq_cells_art <- df_uniq_cells %>%
  dplyr::filter(exp == "ART")

df_uniq_cells_acute <- df_uniq_cells %>%
  dplyr::filter(exp == "Acute")


df_uniq_cells_chronic <- df_uniq_cells %>%
  dplyr::filter(exp == "Chronic")


cell_colors_art <- brewer.pal(length(df_uniq_cells_art$sample_cell), "Paired")
cell_colors_acute <- brewer.pal(length(df_uniq_cells_acute$sample_cell), "Paired")
cell_colors_chronic <- brewer.pal(length(df_uniq_cells_chronic$sample_cell), "Paired")

arrange_posID <- function(posID) {
  set.seed(1)
  return(sample(posID))
}

for (p in uniq_patients) {
  if (grepl("A_", p)) {
    cell_colors <- cell_colors_acute
    uniq_cells <- df_uniq_cells_acute$sample_cell
    abund_threshold <- 3
    
  } else if (grepl("C_", p )) {
    cell_colors <- cell_colors_chronic
    uniq_cells <- df_uniq_cells_chronic$sample_cell
    abund_threshold <- 2
    
  } else {
    cell_colors <- cell_colors_art
    uniq_cells <- df_uniq_cells_art$sample_cell
    abund_threshold <- 2
  }
  
  df_chord_test <- df_compiled %>%
    mutate(sample_cell = factor(paste(location, cell, sep = "_"),
                                levels = uniq_cells)) %>%
    dplyr::filter(patient_anon == p) %>%
    ungroup() %>%
    arrange(sample_cell) %>%
    group_by(sample_cell) %>%
    arrange(arrange_posID(posID), .by_group = TRUE) %>%
    add_tally(name = "group_n") %>%
    ungroup() %>%
    mutate(posID = factor(posID, levels = unique(posID)),
           from_posID = seq_along(posID),
           uniq_posID = as.numeric(posID),
           value = ifelse(duplicated(posID) == TRUE, 1, 0),
           cell_track_color = ifelse(group_n == 1,
                                     paste0(cell_colors[sample_cell], "60"),
                                     "#FFFFFFFF"),
           abund_color = ifelse(abund >= abund_threshold, "#000000", "#CCCCCC"))
    
  # png(paste("figs/chord_", p, "_new.png", sep = ""),
  #     width = 1800, height = 1800, bg = "transparent")
  
  svglite(paste("figs/chord_", p, "_new.svg", sep = ""),
          bg = "transparent",
          system_fonts = list(sans = "Arial"),
          width = 10,
          height = 10)
  draw_chord(df_chord_test, uniq_cells, cell_colors, abund_scale)
  dev.off()
}

make_legend <- function(cell_colors, uniq_cells, fn) {
  df_legend <- data.frame(cell_col = cell_colors,
              cell = factor(uniq_cells, levels = uniq_cells),
              val = rnorm(length(uniq_cells)))

  g <- ggplot(df_legend, aes(x = uniq_cells, y = val, fill = cell)) +
    geom_bar(stat = "identity", color = "#FFFFFF") +
    scale_fill_manual(values = paste(cell_colors, "80", sep = "")) +
    theme(legend.position = "bottom",
          legend.text = element_text(size = 12),
          panel.grid = element_blank(),
          panel.background = element_blank(),
          axis.line = element_blank(),
          axis.title = element_blank(),
          axis.text = element_blank(),
          axis.ticks = element_blank(),
          plot.background = element_blank(),
          legend.title = element_blank()) +
    guides(fill = guide_legend(override.aes = list(size = 10)))
  
  # leg <- cowplot::get_legend(g)
  # grid.newpage()
  # grid.draw(leg)
  ggsave(fn, plot = g, device = "svg", width = 9)
}

make_legend(cell_colors_art,
            df_uniq_cells_art$sample_cell,
            "figs/chord_art_legend.svg")

make_legend(cell_colors_acute,
            df_uniq_cells_acute$sample_cell,
            "figs/chord_acute_legend.svg")

make_legend(cell_colors_chronic,
            df_uniq_cells_chronic$sample_cell,
            "figs/chord_chronic_legend.svg")

```

# Genomic Heatmaps
```{r}
convert_to_mrc_acceptable <- function(df) {
  df_new <- df %>%
    uncount(floor(abund)) %>%
    group_by(sampleName) %>%
    dplyr::mutate(group_n = n()) %>%
    dplyr::filter(group_n >= 30) %>%
    dplyr::ungroup() %>%
    arrange(sampleName) %>%
    dplyr::mutate(siteID = row_number(),
                  gender = "m") %>%
    data.frame()
  
  return(df_new)
}

make_heatmaps <- function(df, annot_ref, filter_type) {
  heatmap_dir <- paste0("heatmaps/", filter_type)
  heatmap_genomic <- paste0(heatmap_dir, "_genomic")
  heatmap_epigenetic <- paste0(heatmap_dir, "_epigenetic")
  
  df_sites_meta <- convert_to_mrc_acceptable(df)
  sites_mrcs <- generate_mrcs(df_sites_meta)
  
  # make genomic
  gen_roc_fn <- file.path(heatmap_genomic, "roc.res.rds")
  if (!file.exists(gen_roc_fn)) {
    print('making genomic annotations')
    
    sites_mrcs_genomic <- annotate_genomic(sites_mrcs, annot_ref)
    sites_to_ROC_ordinary(sites_mrcs_genomic,
                          df_sites_meta[, c("siteID", "sampleName", "refGenome")],
                          heatmap_genomic)
  }
  
  print('making genomic heatmap')
  gen <- heatmap_to_ggplot2(gen_roc_fn, "genomic", NULL, filter_type)
  
  # make epigenetic
  epi_roc_fn <- file.path(heatmap_epigenetic, "roc.res.rds")
  histoneorder <- epigenetic_features()
  histoneorder <- histoneorder[!grepl("[-_]", histoneorder)]
  histoneorder <- histoneorder[!grepl("Nucleo", histoneorder)]
  histoneorder <- str_sort(histoneorder, numeric = TRUE)
  
  if (!file.exists(epi_roc_fn)) {
    print('making epigenetic annotations')
    

    
    sites_mrcs_epigenetic <- annotate_epigenetic(sites_mrcs, annot_ref, histoneorder)
    sites_to_ROC_ordinary(sites_mrcs_epigenetic,
                          df_sites_meta[, c("siteID", "sampleName", "refGenome")],
                          heatmap_epigenetic)
  }
  
  print('making epigenetic heatmap')
  epi <- heatmap_to_ggplot2(epi_roc_fn, "epigenetic", histoneorder, filter_type)
  
  print(gen)
  print(epi)
  
  return("finished")
}

# settings for heatmaps and files
annot_ref <- file.path(normalizePath("annot_ref"))

mark_clonal <- function(exp, abund) {
  clonal <- rep("single", length(abund))
  
  clonal[exp == "Acute" & abund >= 3] <- "clonal"
  clonal[exp == "ART" & abund >= 2] <- "clonal"
  clonal[exp == "Chronic" & abund >= 2] <- "clonal"
  
  return(clonal)
}

df_pre_mrc <- df_compiled %>%
  mutate(refGenome = "hg38",
         type = "insertion",
         clonal = mark_clonal(exp, abund),
         position = as.numeric(str_extract(posID, "(?<=[+-])(\\d*$)")))
```

```{r}
# exp and location and cell
df_pre_mrc %>%
  mutate(sampleName = paste(exp, location, cell, sep = "_")) %>%
  make_heatmaps(annot_ref, "exp_location_cell")

# for (i in seq_along(rangesToCalc)) {
#   row <- test[i]
#   s <- as.character(seqnames(row))
#   pos <- row@ranges@start + row@ranges@width
# 
#   if (is.na(pos) | pos >= seqlengths(hg38)[s]) {
#     print(row)
#   }
# }


# hg38 <- get("BSgenome.Hsapiens.UCSC.hg38")
# library(BSgenome)
# library(GCcontent)
# 
# window_size_GC <- c("100"=100, 
#                       "1k"=1000, "10k"=1e4, "100k"=1e5, "1M"=1e6)
# 
# 
# sites <- test
# nsites <- length(sites)
# strand(sites) = "+"
# sites.seqinfo.original <- seqinfo(sites)
# isCircular(seqinfo(sites)) <- rep(FALSE, length(seqinfo(sites)))
# 
# sites <- rep(sites, length(window_size_GC))
# sites <- GenomicRanges::trim(suppressWarnings(flank(sites, rep(window_size_GC/2, 
#   each = nsites), both = T)))
# seqinfo(sites) = sites.seqinfo.original
# sites
# 
# 
# rangesToCalc <- GCcontent:::.expand_trim_GRanges(test, window_size_GC)
# seqs <- getSeq(hg38, sites, as.character = F)
# 

# exp and location and clonal
df_pre_mrc %>%
  mutate(sampleName = paste(exp, location, clonal, sep = "_")) %>%
  make_heatmaps(annot_ref, "exp_location_clonal")

# art_only and location and cell
df_pre_mrc %>%
  filter(exp == "ART") %>%
  mutate(sampleName = paste(location, cell, sep = "_")) %>%
  make_heatmaps(annot_ref, "ARTonly_location_cell")

# acute_only and location and cell
df_pre_mrc %>%
  filter(exp == "Acute") %>%
  mutate(sampleName = paste(location, cell, sep = "_")) %>%
  make_heatmaps(annot_ref, "Acuteonly_location_cell")

# acute_only and location and cell
df_pre_mrc %>%
  filter(exp == "Chronic") %>%
  mutate(sampleName = paste(location, cell, sep = "_")) %>%
  make_heatmaps(annot_ref, "Chroniconly_location_cell")
```

# overall stat counts
```{r}
df_compiled %>%
  group_by(exp) %>%
  summarise(count = n(),
            abund = sum(abund))

df_compiled %>%
  group_by(exp) %>%
  arrange(abund) %>%
  top_n(5, abund) %>%
  dplyr::select(exp, abund, gene_plot, in_gene)
```

# site annotations
```{r}
covplot(gr_compiled, title = "Integration site distribution over chromosome")
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene

gr_anno <- annotatePeak(gr_compiled, TxDb = txdb, annoDb = "org.Hs.eg.db", overlap = "all")

# uniq_cells <- unique(gr_compiled$cell)
# for (cell in uniq_cells) {
#   gr_subset <- gr_compiled[gr_compiled$cell == cell]
#   if (length(gr_subset < 10)) {
#     next
#   }
#   
#   print(cell)
#   gr_subset_anno <- annotatePeak(gr_subset,
#                             TxDb = txdb,
#                             annoDb = "org.Hs.eg.db",
#                             overlap = "all",
#                             verbose = FALSE)
#   
#   pdf(paste0("figs/upsetplot_", cell, ".pdf"), width = 6, height = 4)
#   print(upsetplot(gr_subset_anno))
#   dev.off()
# }

pdf(paste0("figs/upsetplot_", 'general', ".pdf"), onefile = FALSE, width = 6, height = 4)
upsetplot(gr_anno)
dev.off()

df_anno <- data.frame(gr_anno) %>%
  mutate(feature = case_when(grepl("Intron", annotation) ~ "Intron",
    grepl("Distal Intergenic", annotation) ~ "Distal intergenic",
    grepl("Exon", annotation) ~ "Exon",
    TRUE ~ "Other")) %>%
  mutate(feature = factor(feature, levels = c("Intron", "Distal intergenic", "Exon", "Other"))) %>%
  group_by(exp, location, cell) %>%
  mutate(group_n = sum(abund)) %>%
  group_by(exp, location, cell, feature) %>%
  summarise(percent_feature = sum(abund) / max(group_n) * 100,
            sum_feature = sum(abund),
            group_abund = max(group_n)) %>%
  mutate(x_axis = paste0(gsub(" ", "\n", cell),
                         "\n(",
                         group_abund,
                         ")")) %>%
  ungroup() %>%
  mutate(exp = factor(exp, levels = c("Acute", "Chronic", "ART")),
         location = factor(location, levels = c("PBMC", "CLN", "ILN", "Tonsil")))


# chi square test to test cell subset influence on the genomic site proportions
# df_anno %>%
#   ungroup() %>%
#   dplyr::filter(exp == "ART-treated", location == "Tonsil") %>%
#   dplyr::select(cell, feature, sum_feature) %>%
#   pivot_wider(names_from = feature,
#               values_from = sum_feature) %>%
#   dplyr::filter(!cell == "Naive") %>%
#   dplyr::select(-cell) %>%
#   chisq.test(., simulate.p.value = TRUE)


facet_label_clean <- function(labels) {
  labels <- lapply(labels, as.character)
  a <- do.call(paste, c(labels, list(sep = " | ")))
  
  return(list(a))
}

ggplot(df_anno, aes(x = x_axis, y = percent_feature, fill = feature)) +
  geom_col(position = position_dodge2(preserve = "single")) +
  labs(fill = "Feature",
       y = "% sonic abundance",
       x = "Cell subset\n(total sonic abundance)") +
  theme_classic() +
  theme(strip.background = element_rect(fill = "#FFFFFF", colour = "#FFFFFF"),
        strip.text = element_text(face = "bold", size = 10),
        legend.title = element_text(size = 10),
        legend.text = element_text(size = 8),
        axis.text.x = element_text(size = 8)) +
  guides(color = guide_legend(override.aes = list(size = 0.75))) +
  facet_wrap(~ exp + location,
             scales = "free_x",
             labeller = facet_label_clean)

ggsave(filename = "figs/site_annot_subset.svg", device = "svg", width = 9, height = 6)
```

# in_gene comparison
```{r}
refGenes <- readRDS("reference_hg38.rds")
refGenes <- refGenes[
  seqnames(refGenes) %in% paste0("chr", c(1:22, "X", "Y"))
]
mod_refGenes <- unlist(GenomicRanges::reduce(split(
    refGenes, refGenes$name2), min.gapwidth = 100000L))
mod_refGenes$name <- names(mod_refGenes)

get_loci_id <- function(gr){
  paste0(seqnames(gr), strand(gr), start(gr), ":", end(gr))
}

all_in_gene <- findOverlaps(mod_refGenes, gr_compiled, maxgap = 5000, ignore.strand = TRUE)

art_patients <- c("A", "B", "C")
acute_patients <- c("A_D", "A_E", "A_F")

all_stats <- as.data.frame(gr_compiled[subjectHits(all_in_gene)], row.names = NULL) %>%
  mutate(loci = get_loci_id(mod_refGenes[queryHits(all_in_gene)]),
    gene_name = as.character(mod_refGenes$name[queryHits(all_in_gene)]),
    gene_ort = as.character(strand(mod_refGenes[queryHits(all_in_gene)])),
    gene_width = (width(mod_refGenes$name[queryHits(all_in_gene)]) + 10000) / 1000,
    strand = as.character(strand)) %>%
  dplyr::select(loci, gene_name, gene_ort, posID, gene_width, strand, abund, exp, patient_anon) %>%
  group_by(loci, gene_name, gene_ort, gene_width) %>%
  summarise(n_patients = n_distinct(patient_anon),
            acute_n_patients = n_distinct(patient_anon[patient_anon %in% acute_patients]),
            art_n_patients = n_distinct(patient_anon[patient_anon %in% art_patients]),
            n_sites = n_distinct(posID),
            acute_n_sites = n_distinct(posID[patient_anon %in% acute_patients]),
            art_n_sites = n_distinct(posID[patient_anon %in% art_patients]))


acute_df <- df_compiled %>% filter(exp == "Acute")
art_df <- df_compiled %>% filter(exp == "ART-treated")

total_acute_sites <- length(unique(paste(acute_df$patient_anon, acute_df$posID)))
total_art_sites <- length(unique(paste(art_df$patient_anon, art_df$posID)))
total_pat_sites <- length(unique(paste(df_compiled$patient, df_compiled$posID)))

all_stats %>%
  dplyr::filter(n_patients > 0) %>%
  ungroup() %>%
  mutate(acute_freq = acute_n_sites / (gene_width * total_acute_sites),
         art_freq = art_n_sites / (gene_width * total_art_sites),
         freq_diff = (art_freq - acute_freq),
         pct_chg = 100 * freq_diff / acute_freq) %>%
  dplyr::filter(art_n_patients >= 1, art_n_sites >= 2, acute_n_sites > 0) %>%
  arrange(desc(pct_chg)) %>%
  dplyr::select(gene_name, acute_n_patients, art_n_patients,
                acute_n_sites, art_n_sites, acute_freq, art_freq, pct_chg) %>%
  write.csv(x = ., file = "site_enrichment.csv", row.names = FALSE, quote = FALSE)
  # dplyr::mutate(id = seq_len(n())) %>%
  # dplyr::group_by(id) %>%
  # dplyr::mutate(freq_fisher_test = fisher.test(
  #   matrix(
  #     c(acute_n_sites, art_n_sites, total_acute_sites, total_pat_sites), 
  #     ncol = 2))$p.value) %>%
  # ungroup() %>%
  # dplyr::mutate(adj_freq_test = p.adjust(freq_fisher_test, method = "BH"))
```

# look at position of integration site on chromosome
```{r}
find_nn_dist <- function(start) {
  nn_dist <- sapply(seq_along(start), function(i) {
    if (i == 1) {
      neighbor <- start[i + 1]
    } else if (i == length(start)) {
      neighbor <- start[length(start) - 1]
    } else {
      pre <- start[i + 1]
      post <- start[i - 1]
      
      neighbor <- ifelse(abs(pre - start[i]) < abs(post - start[i]),
                         pre,
                         post)
    }
    return(abs(start[i] - neighbor))
  })
  
  return(nn_dist)
}

df_compiled %>%
  dplyr::filter(exp == "ART") %>%
  group_by(patient_anon, seqnames) %>%
  arrange(start, .by_group = TRUE) %>%
  mutate(neighbor = find_nn_dist(start)) %>%
  dplyr::filter(neighbor < 100) %>%
  ggplot(aes(x = neighbor)) +
  stat_ecdf(geom = "step") +
  facet_wrap(~ seqnames, scales = "free") +
  theme_classic()

# TODO come up with graph...
hg38 <- readRDS("reference_hg38.rds")
```

