---
title: "hiv_intsite_compiled"
author: "Vincent Wu"
date: "10/3/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# load pkgs
```{r, warning=FALSE, message=FALSE}
library(ggplot2)
library(tidyverse)
library(GenomicRanges)
library(hiAnnotator)
library(RColorBrewer)
library(scales)
library(circlize)
library(grid)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
library(org.Hs.eg.db)
library(ChIPseeker)
library(BSgenome.Hsapiens.UCSC.hg38)
library(hotROCs)

source("scripts/hiv_int_helper_fx.R")
source("scripts/random_site.R")
```

# load and merge data files
```{r}
df_art <- readRDS("data/df_compiled_1-5.rds")
df_acute <- readRDS("data/df_compiled_6.rds")

df_compiled <- df_art %>%
  mutate(exp = "ART-treated") %>%
  bind_rows(df_acute %>% mutate(exp = "Acute"))

# NOTE: removing chrM sample, likely mappinga artifact.
df_compiled <- df_compiled %>%
  filter(seqnames != "chrM")

# NOTE: remove patients D/E (no paired pbmc)
df_compiled <- df_compiled %>%
  filter(!patient_anon %in% c("D", "E"))

df_compiled %>%
  dplyr::select(-patient) %>%
  saveRDS(file = "data/df_compiled_vincent.rds")

gr_compiled <- makeGRangesFromDataFrame(df_compiled,
                                        keep.extra.columns = TRUE)
```

# patient-wide simulations
```{r}
calc_overlap <- function(ids, bin_index) {
  abund_ids <- ids[duplicated(ids)]
  total_count <- 0
  
  for (i in abund_ids) {
    bin_count <- 0
    for (b in seq(2, length(bin_index) + 1, 1)) {
      start <- bin_index[b - 1]
      
      if (b == length(bin_index) + 1) {
        end <- length(ids)
      } else {
        end <- bin_index[b] - 1        
      }
      
      bin <- ids[start:end]
      if (sum(bin == i) > 0) {
        bin_count <- bin_count + 1
      }
    }
    
    if (bin_count >= 2) {
      total_count <- total_count + 1
    }
  }
  
  return(total_count)
}

uniq_patients <- unique(df_compiled$patient_anon)
graphs <- list()

for (p in uniq_patients) {
  print(paste("Running simulation for", p))
  
  trial_df <- df_compiled %>%
    filter(patient_anon == p) %>%
    mutate(location_cell = paste(location, cell, sep = "_")) %>%
    mutate(group_id = group_indices(., location_cell)) %>%
    arrange(group_id)
  
  bin_index <- which(!duplicated(trial_df$group_id))
  ids <- trial_df$posID
  weights <- trial_df$abund
  
  # make pool with correct abundances
  pool_ids <- rep(ids, times = floor(weights))
  
  # find weights by group
  g_sum_df <- trial_df %>%
    group_by(group_id) %>%
    summarise(g_sum = sum(floor(abund)))
  
  weights_grouped <- g_sum_df$g_sum
  
  # get expanded bins (bins that account for abundances)
  # bin_index_exp <- c(1, weights_grouped[1])
  # for (i in seq(2, length(weights_grouped) - 1)) {
  #   w <- weights_grouped[i]
  #   
  #   bin_index_exp <- c(bin_index_exp, bin_index_exp[length(bin_index_exp)] + w + 1)
  # }
  
  bin_index_exp <- cumsum(weights_grouped)
  bin_index_exp <- bin_index_exp[-length(bin_index_exp)] + 1
  bin_index_exp <- c(1, bin_index_exp)
  
  # calculate overlap
  act_overlaps <- sum(duplicated(ids))
  
  # run simulations for overlap
  test <- sapply(seq(1, 1000), function(x) {
    pool_scramble <- sample(pool_ids)
    calc_overlap(pool_scramble, bin_index_exp)
  })
  
  weights_grouped
  
  # print(paste("Actual overlaps:", act_overlaps))
  
  # clone_info <- trial_df %>% filter(abund > 1)
  # print(summary(clone_info$abund))
  
  g_df <-  data.frame(x = test)
  
  g <- ggplot(data = g_df, aes(x = test)) +
    # geom_histogram(fill = "#CCCCCC", color = "#AAAAAA") +
    geom_density(fill = "#CCCCCC", color = "#AAAAAA") +
    theme_classic() +
    labs(title = p)
  
  graphs[[p]] <- g
    
  ggsave(filename = paste("figs/global_overlap_dens_", p, ".png", sep = ""),
         plot = graphs[[p]],
         device = "png")
}

# multiplot(plotlist = graphs, cols = 2)
```

# compile dataset from published sources
```{r}
df <- read.csv("data/20191004_hiris.csv", header = TRUE, stringsAsFactors = FALSE) %>%
  filter(environment == "in vivo")

pmid_of_interest <- c("15163705", "17262715", "24968937", "25011556")

df_interest <- df %>%
  filter(pubmed_ids %in% pmid_of_interest) %>%
  mutate(strand = ifelse(orientation_in_landmark == "", "*",
                         ifelse(orientation_in_landmark == "F", "+", "-")))

gr_interest <- makeGRangesFromDataFrame(df_interest,
                                       keep.extra.columns = TRUE,
                                       seqnames.field = "landmark",
                                       start.field = "location",
                                       end.field = "location",
                                       strand.field = "strand")

gr_interest

ref <- make_ref("reference_hg38.rds")
gr_interest <- getSitesInFeature(gr_interest, ref, "in_gene")
gr_interest <- getNearestFeature(gr_interest, ref, "nearest_gene")

df_clonal <- data.frame(gr_interest) %>%
  filter(in_gene != "FALSE") %>%
  group_by(in_gene) %>%
  tally() %>%
  arrange(desc(n))

write.table(df_clonal,
            "historical_art_intsites.tsv",
            sep = "\t",
            row.names = FALSE,
            col.names = FALSE,
            quote = FALSE)
```

# Try getting MRCs
```{r}
df_art <- df_compiled %>%
  filter(exp == "ART-treated") %>%
  mutate(refGenome = "hg38",
         type = "insertion",
         sampleName = paste(patient, cell, location, sep = "_"),
         position = as.numeric(str_extract(posID, "(?<=[+-])(\\d*$)")))

df_sites_meta <- df_art %>%
  group_by(sampleName) %>%
  dplyr::mutate(group_n = n()) %>%
  dplyr::filter(group_n >= 10) %>%
  dplyr::ungroup() %>%
  arrange(sampleName) %>%
  mutate(siteID = row_number(),
         gender = "m") %>%
  data.frame()
  # filter(sampleName %in% c("MAR_CD69+_Tonsil", "GMCB_Tem_PBMC", "MAR_CD69-_Tonsil", "MAR_Tem_PBMC")) %>%

keepBSgenomeSequences <- function(genome, seqnames) {
    stopifnot(all(seqnames %in% seqnames(genome)))
    genome@user_seqnames <- setNames(seqnames, seqnames)
    genome@seqinfo <- genome@seqinfo[seqnames]
    genome
}

hg38 <- BSgenome.Hsapiens.UCSC.hg38
sequences_to_keep <- paste0("chr", c(1:22, "X", "Y"))
hg38 <- keepBSgenomeSequences(hg38, sequences_to_keep)

mrcs <- get_N_MRCs(df_sites_meta %>% dplyr::select(siteID, gender), hg38)
mrcs <- merge(mrcs, df_sites_meta[,c("siteID", "sampleName", "refGenome")])
mrcs$type <- "match"

mrcs <- merge(mrcs,
              df_sites_meta[,c("siteID", "sampleName", "refGenome", "gender")]) %>%
  dplyr::select(-refGenome)
  

sites <- df_sites_meta %>%
  dplyr::rename(chr = seqnames) %>% 
  dplyr::select(siteID, sampleName, gender, chr, strand, position, type)

sites_mrcs <- rbind(sites, mrcs) %>%
  mutate(position = as.numeric(position))

sites_mrcs <- makeGRanges(sites_mrcs,
                          soloStart=TRUE,
                          chromCol='chr',
                          strandCol='strand',
                          startCol='position')

newSeqInfo <- seqinfo(hg38)
seqInfo.new2old <- match(seqnames(newSeqInfo), seqnames(seqinfo(sites_mrcs)))
seqinfo(sites_mrcs, new2old=seqInfo.new2old) <- newSeqInfo
```

```{r}
source("scripts/genomicHeatmapMaker_vincent.R")
reference_genome <- "hg38"

refseq_rds <- "rds/hg38.refSeq.rds"
if (!file.exists(refseq_rds)) {
  refSeq_genes <- getRefSeq_genes(reference_genome)
  # refSeq_genes <- keepStandardChromosomes(refSeq_genes, pruning.mode = "coarse")
  # seqinfo(refSeq_genes) <- seqinfo(hg38)
  
  saveRDS(refSeq_genes, refseq_rds)
} else {
  refSeq_genes <- readRDS(refseq_rds)
}

cpg_rds <- "rds/hg38.cpg.rds"
if (!file.exists(cpg_rds)) {
  CpG_islands <- get_cpg(reference_genome, hg38)
  
  saveRDS(CpG_islands, cpg_rds)
} else {
  CpG_islands <- readRDS(cpg_rds)
}

oncogenes <- get_oncogene_from_file("scripts/allonco_no_pipes.csv")

stability_regions <- readRDS("scripts/fragileRegions.hg38.rds")

fragile_regions <- stability_regions[
  stability_regions$regionType == "aCFR"]
non_fragile_regions <- stability_regions[
  stability_regions$regionType == "NFR"]

window_size_fragReg <- c("1M"=1e6)

sites_mrcs <- getFeatureCounts(
  sites_mrcs, non_fragile_regions, "NFR", width = window_size_fragReg)
sites_mrcs <- getFeatureCounts(
  sites_mrcs, fragile_regions, "aCFR", width = window_size_fragReg)

# is there oncogene closer than 50k
refSeq_gene_symbols <- refSeq_genes$name2
#' check if gene is onco gene list(curated by Bushman's lab)
#' @return TRUE if onco-gene FALSE if not
is_onco_gene <- function(gene_symbol_sites, oncogenes) {
    toupper(gene_symbol_sites) %in% toupper(oncogenes)
}
is_refSeq_oncogene <- is_onco_gene(refSeq_gene_symbols, oncogenes)
refSeq_oncogene <- refSeq_genes[is_refSeq_oncogene]
sites_mrcs <- getNearestFeature(
  sites_mrcs, refSeq_oncogene, dists.only=TRUE, colnam="onco")
#sites_mrcs, refSeq_oncogene, dists.only=TRUE, colnam="onco.100k")
sites_mrcs$onco.100k <- abs(sites_mrcs$oncoDist) <= 50000
sites_mrcs$oncoDist <- NULL
# end oncogene

# need within_refSeq_gene for getPositionalValuesOfFeature
sites_mrcs <- getSitesInFeature(
  sites_mrcs, refSeq_genes, "within_refSeq_gene", asBool=TRUE)

sites_mrcs <- getPositionalValuesOfFeature(sites_mrcs, refSeq_genes)

# redo the work to move it after positional values 
mcols(sites_mrcs)$within_refSeq_gene <- NULL
sites_mrcs <- getSitesInFeature(
  sites_mrcs, refSeq_genes, "within_refSeq_gene", asBool=TRUE)

window_size_refSeq <- c("10k"=1e4, "100k"=1e5, "1M"=1e6)
sites_mrcs <- getFeatureCounts(sites_mrcs, refSeq_genes, "refSeq_counts", 
                               width=window_size_refSeq)

window_size_GC <- c("100"=100, 
    "1k"=1000, "10k"=1e4, "100k"=1e5, "1M"=1e6)
sites_mrcs <- getGCpercentage(
  sites_mrcs, "GC", window_size_GC, hg38)

window_size_CpG_counts <- c("1k"=1e3, "10k"=1e4)
sites_mrcs <- getFeatureCounts(sites_mrcs, CpG_islands, "CpG_counts", 
                               width=window_size_CpG_counts)

window_size_CpG_density <- c("10k"=1e4, "100k"=1e5, "1M"=1e6)
sites_mrcs <- getFeatureCounts(sites_mrcs, CpG_islands, "CpG_density", 
                               width=window_size_CpG_density)
sites_mrcs <- from_counts_to_density(sites_mrcs, 
                                     "CpG_density", window_size_CpG_density)


dnase_rds <- "rds/hg38.dnase.rds"
if (!file.exists(dnase_rds)) {
  DNaseI_data <- getUCSCtable(
    "wgEncodeRegDnaseClustered",
    "DNase Clusters",
    freeze = reference_genome)
          
  DNaseI_data <- DNaseI_data[
    DNaseI_data$chrom %in% 
      seqnames(GenomeInfoDb::seqinfo(hg38)),
    ]
  
  DNaseI <- GenomicRanges::GRanges(
    seqnames = DNaseI_data$chrom,
    ranges = IRanges::IRanges(
      start = DNaseI_data$chromStart, end = DNaseI_data$chromEnd
    ),
    strand = "*",
    seqinfo = GenomeInfoDb::seqinfo(hg38)
  )
  
  mcols(DNaseI) <- DNaseI_data
  
  saveRDS(DNaseI, dnase_rds)
} else {
  DNaseI <- readRDS(dnase_rds)
}

window_size_DNaseI <- c("1k"=1e3, "10k"=1e4, "100k"=1e5, "1M"=1e6)
sites_mrcs <- getFeatureCounts(sites_mrcs, DNaseI, "DNaseI_count", 
                               width=window_size_DNaseI)

sites_to_ROC_ordinary(sites_mrcs,
                      df_sites_meta[, c("siteID", "sampleName", "refGenome")],
                      "heatmaps")
```


